src/metatrain/experimental/soap_bpnn/spherical.py

```diff
@@ -0,0 +1,401 @@
+"""Modules to allow SOAP-BPNN to fit arbitrary spherical tensor targets."""
+
+import copy
+from typing import Dict, List, Optional
+
+import metatensor.torch
+import numpy as np
+import rascaline.torch
+import torch
+import wigners
+from metatensor.torch import Labels
+from metatensor.torch.atomistic import System
+from metatensor.torch.learn.nn import Linear as LinearMap
+
+
+class VectorBasis(torch.nn.Module):
+    """
+    This module creates a basis of 3 vectors for each atomic environment.
+
+    In practice, this is done by contracting a l=1 spherical expansion.
+    """
+
+    def __init__(self, atomic_types, soap_hypers) -> None:
+        super().__init__()
+        self.atomic_types = atomic_types
+        soap_vector_hypers = copy.deepcopy(soap_hypers)
+        soap_vector_hypers["max_angular"] = 1
+        self.soap_calculator = rascaline.torch.SphericalExpansion(
+            radial_basis={"Gto": {}}, **soap_vector_hypers
+        )
+        self.neighbor_species_labels = Labels(
+            names=["neighbor_type"],
+            values=torch.tensor(self.atomic_types).reshape(-1, 1),
+        )
+        self.contraction = LinearMap(
+            in_keys=Labels(
+                names=["o3_lambda", "o3_sigma", "center_type"],
+                values=torch.stack(
+                    [
+                        torch.tensor([1] * len(self.atomic_types)),
+                        torch.tensor([1] * len(self.atomic_types)),
+                        torch.tensor(self.atomic_types),
+                    ],
+                    dim=1,
+                ),
+            ),
+            in_features=soap_vector_hypers["max_radial"] * len(self.atomic_types),
+            out_features=3,
+            bias=False,
+            out_properties=[Labels.range("basis", 3) for _ in self.atomic_types],
+        )
+        # this optimizable basis seems to work much better than a fixed one
+
+    def forward(
+        self, systems: List[System], selected_atoms: Optional[Labels]
+    ) -> torch.Tensor:
+        device = systems[0].positions.device
+        if self.neighbor_species_labels.device != device:
+            self.neighbor_species_labels = self.neighbor_species_labels.to(device)
+
+        spherical_expansion = self.soap_calculator(
+            systems, selected_samples=selected_atoms
+        )
+
+        # by calling keys_to_samples and keys_to_properties in the same order as they
+        # are called in the main model, we should ensure that the order of the samples
```

Comment:
Alternatively, you could pass an explicit argument to make sure the same keys are moved in the same order

Reply:
What argument?

Reply:
`keys_to_move`: https://docs.metatensor.org/latest/torch/reference/tensor.html#metatensor.torch.TensorMap.keys_to_samples

Reply:
Ok, thanks, I think it's much easier to do what we're doing now